### ISSUE TITLE
Fails to update MFA Settings

### ISSUE THREAD

**Issue by dsielert** [aws-quickstart/quickstart-duo-mfa#10]
related exactly to #1 

Found addresses: [], but two are required.

From the lambda function. 

authproxy log is empty
there are no other useful information in the logs and in the troubleshooting guide. 

I've deleted and recreated this template 4 times now with the same result.

**Comment by dsielert** [comment]
it seems as though these tags the instances are filtered by are never applied to the instances
     {"Name": "tag:duo:DirectoryServiceId", "Values": [tagvalue]},
            {"Name": "tag:RadiusConfigured", "Values": ["True"]},
            {"Name": "instance-state-name", "Values": ["running"]}

**Comment by dsielert** [comment]
specifically the last two RadiusConfigured and instance-state-name are missing from the created instances

**Comment by toytown** [comment]
I am also having same issue, after deployment. This is the second time and it is not working. Troubleshooting does not help.  The MFA part is not updated in the Directory .  I had the same issue "Found addresses: [], but two are required." but resolved it after adding DNS port 53 in the DuSecurity group described in one o the issue here. Even with this I am getting error below.  I even tried to update the MFA setting in the Directory manually but it did not work.

![image](https://user-images.githubusercontent.com/535868/74907012-70309800-53aa-11ea-9e19-8ce0a63ebb6a.png)

**Comment by dsielert** [comment]
I believe I solved this issue by opening the security groups of the AD, buried in the template I believe it's related to how to finds the AD ip's

**Comment by toytown** [comment]
Thanks. Could you provide what specific changes you made to AD's security group like port number etc.

**Comment by dsielert** [comment]
I believe you just have to open up 389/689

**Comment by toytown** [comment]
Still struggling to get this running. Those ports are part of security group already (inbound rule). Also I do not see any VPC logs. Are you using Managed AD and did you made some changes like Firewall or NPS settings ?

**Comment by gatesman466** [comment]
Having the same problem. To try to resolve the issue, using the suggestion from dsielert, I opened Inbound TCP\UDP 389 & 689 from IPv4 0/0 on the SG used by the domain controllers in my AWS Managed AD. That didn't work. 

Next I opened Inbound All Traffic from IPv4 0/0 on the same SG. That worked. Surprise, surprise. So dsielert in on the right path with opening the SG to allow for proper IP address discovery, but leaving inbound All Traffic from 0/0 isn't a long term solution. Next I'll try to winnow it down to just the needed ports.

**Comment by barnesrobert** [comment]
Hi, gatesman and toytown. 

The egress ports on the RADIUS security group need to have 80 and 443 open to the internet so they can download the Duo authproxy service and communicate to the Systems Manager service. They also need UDP port 1812 to the domain controller security group. By default, the Directory Service security group allows ingress from 0/0, but if you modified the rules, you would need to ensure that those allow ingress from the RADIUS security group.

One other special case is that If you're using Systems Manager VPC endpoints, then you'll need to ensure that you open 443 to the security group associated with those.

**Comment by gatesman466** [comment]
Tested this more extensively to find the Directory Service SG rules needed to to allow updating of the Directory MFA Settings. Here's what I found. 

When using AWS Managed AD, the default Inbound rules for the Directory Service SG allow many ports from 0/0. However, the rules do not allow Inbound TCP\UDP 1812 from 0/0. You need to add this. Port 1812 is the network access Lambda needs in order to update the MFA settings on the Managed AD Directory. 

Further, when you add that rule, it's not effective to all inbound traffic from the RADIUS security group. If you need to delete and re-run the Quickstat CFN template, that Security Group ID will change, and the SG rule you created will no longer be valid. I avoided this problem by allowing Inbound 1812 from 0/0. Not ideal, but since nearly all the other Inbound rules allowed traffic from 0/0, at least it was in keeping with that. 

Finally, a note on troubleshooting. After running this CFN a couple dozen times, I discovered that the Directory MFA settings aren't updated by the time the CFN reaches the "CREATE_COMPLETE" state. It takes longer. Usually, it takes about 5-10 minutes more. 

Why? Because the Lambda function that updates the Directory MFA settings is checking Cloudwatch Events to discover (a) when the RADIUS instances are done launching and (b) their IP addresses. In my experience, it takes 3 cycles of checks for Lambda to discover both IP addresses, and then move forward with the MFA updating process. 

If you look in the Cloudwatch Log group for UpdateDirectoryServiceMfaSettings, and see that Lambda has discovered the IP addresses of your RADIUS instances, but fails at configuring the RADIUS settings of your Directory, add the inbound rules for port 1812 mentioned above.

**Comment by radu-mymed** [comment]
After fighting with this a few times during the past month I have created the following checklist based on starting from scratch, but which can be used when some of the resources already exist as well. I don't know what the best place for this is, so I'm posting it here as a starting point.

Notice that I'm not doing any of the following from the guide:

- creating the app URL
- creating the IAM role
- enabling the AWS Management Console

I'm using this setup with WorkSpaces and none of the above are necessary as far as I can tell. WorkSpaces are outside the scope of this checklist.

# Checklist

## Duo subscription

As this is essential, first get a Duo plan that supports 2FA. The free plan does not. If you get a trial make sure you remember when it expires, because there's no error message telling you about it when you're moved back to the free plan.

## VPC

We need three /24 subnets: one public, two private. We'll use a /22 network for the VPC itself. For this example we'll use 192.168.0.0/22

* CIDR: 192.168.0.0/22
* Subnets:
    - public: 192.168.0.0/24
    - private 1: 192.168.1.0/24
    - private 2: 192.168.2.0/24
* Internet Gateway - create, and attach to VPC
* NAT Gateway - create on the public subnet
* Routing tables:
    - Use the routing table created by the VPC as the public subnet routing table
    - Create two more routing tables for the private subnets
    - public: add route to 0.0.0.0/0 via the Internet Gateway
    - private: add route to 0.0.0.0/0 via the NAT Gateway to both tables

## Directory Services

- Create MS AD on VPC using both private subnets created earlier. This took about 25 minutes the two times I timed it.

## Duo

Worth repeating: make sure you're on a plan that supports 2FA. If you're not (e.g. trial expired) then you won't be told by anything that's the reason for MFA setup failure.

- Create protected application of type RADIUS (one word, without any qualified words around it)

## Quick Start

- Go to the Quick Start page: https://aws.amazon.com/quickstart/architecture/duo-mfa/
- open tab "How to deploy"
- click "Launch the Quick Start"

- Fill in the details from the RADIUS application on Duo and the AD ID.
    Leave everything else as the defaults.
    The bucket defaults may not match your region, but if you don't change those fields than the solution will select the correct region for you.

## Security groups

The security groups currently need some manual massaging.

There are two security groups that you have to touch:
- The security group created by the Quick Start (QS SG)
- The _controllers security group for your AD (_controllers SG)

- QS SG: Outbound: All Traffic to the _controllers SG
- _controllers SG: Outbound: Custom UDP, port 1812, to QS SG

## Systems Manager

- Open the State Manager
- Open the relevant BootstrapRadiusProxies-(AD ID) Association
- Press "Apply Association" (this tiggers the Lambda chain to setup MFA)

## Directory Services

The AD's "Networking and security" tab, "MFA" section, should go into "Creating" mode and after a short while switch to "Completed"

## Conclusion

The most common failures I've seen are from either

- Bad VPC configuration: make sure the subnets, gateways, and routing tables are setup correctly
- The Quick Start is not setting up everything needed in the security groups
- Unpaid Duo plan (trial expired)

When the checklist fails on an existing setup just delete the CloudFormation stack and start over. And if that fails, recreate the protected RADIUS on Duo and make a new one and start over.

**Comment by schemert15** [comment]
Trying to implement this, but it doesnt work.
No logging gets created in the authproxy.log

paid duo mfa subscription
ADConnector in AWS does work correctly with O365, can logon to workspace/workdocs when mfa is disabled

radius servers are in private subnets, connected to pubsubnet with NAT

ADconnector SG:
inbound: all tcp of Radius servers SG
outbound: all allowed

Radiusservers SG
inbound: udp 1812 from ADConnector SG
outbound: tcp 80 and 443 to internet


Can you assist in getting it working?

Thanks,
Sander van Gemert

**Comment by radu-mymed** [comment]
> Trying to implement this, but it doesnt work.
> No logging gets created in the authproxy.log

You make it sounds like you got MFA enabled (it says `Completed` in the AD's Networking tab). Is that correct?

Otherwise, check this log out too: `UpdateDirectoryServiceMfaSettings-d-12345678`, where `d-12345678` is supposed to be your AD ID.

If you don't see two servers discovered, then your Security Groups need massaging.

If you do... then I got nothing. I had to delete my stack and start over and then it worked as per my checklist, but I have a managed AWS AD, not AD Connector.

**Comment by RandyBT** [comment]
Anyone have any luck with this? I allowed any any on everything however I am still seeing deployment issues - Support advised I start an aws ticket..

**Comment by schemert15** [comment]
Yes ours works great. Just delete all and start over again. 
Make sure you have a paid subscription at duo

**Comment by RandyBT** [comment]
I tried using my tenant still no result. I am using an MSP subscription. Not sure if that matters? I still see failed on the MFA

**Comment by schemert15** [comment]
Are you sure ad connect works ok?

**Comment by RandyBT** [comment]
Im using Directory services through aws not the ad connector. Are you using the AD connector?

**Comment by schemert15** [comment]
Yes we have the ad connector to o365. 
But if you have the complete aws directory service it should work too, I believe.

At first ours did not work too. After deleting all and rerun the script it did work

**Comment by schemert15** [comment]
Using aws systems Mgr you can login to the vms and check the logging

**Comment by RandyBT** [comment]
Hmm I am a bit nooby with Linux - do you know where the logs sit?

**Comment by radu-mymed** [comment]
New information: It turns out that the authproxy doesn't always start. If you've done my checklist and it doesn't work, then you can connect to the RADIUS proxy instances using SSM (in the browser) and check a few things out:

```bash
[root@ip-10-0-1-84 ~]# ps ax|grep duo
 6302 ?        Sl     0:00 /opt/duoauthproxy/usr/local/bin/python3 /opt/duoauthproxy/usr/local/bin/twistd --pidfile=/opt/duoauthproxy/run/duoauthproxy.pid --python=/opt/duoauthproxy/bin/duoauthproxy.tap --uid=99 --gid=99

[root@ip-10-0-1-84 ~]# netstat -anp | grep :1812
udp        0      0 0.0.0.0:1812                0.0.0.0:*                               6302/python3
```

If you don't have those two entries (i.e. the process and the listening port) things won't work.

You can also have a look with tcpdump:

```bash
yum install tcpdump
tcpdump -n -l -s0 -A port 1812
```

If you see this sort of stuff:

```
23:50:30.842225 IP 10.0.1.84 > 10.0.1.219: ICMP 10.0.1.84 udp port radius unreachable, length 112
```

Then things won't work.

The annoying thing is that after a while the "udp port radius unreachable" message went away and things worked as if by magic.

Restarting the offending instance might help. Terminating it and letting the load balancer create a new one might help a bit more.

Me? I'd create an AMI from a working Duo Radius proxy and reconfigure the load balancer to start with it, and use it in future deployments. This solution compiles the software in place, and that may be the cause of some of the problems.

**Comment by radu-mymed** [comment]
I've created an AMI from an instance that set up successfully (verified with the steps in my previous comment), set that in a Systems Manager -> Parameter Store value, used that value in the quickstart, and I didn't even need to alter the Security Groups or retrigger the lambdas. It worked first time, all by itself.

I'm guessing that compiling everything from source when a new RADIUS proxy instance starts isn't such a good idea.

**Comment by RandyBT** [comment]
I tore down everything the CloudFormation template re-created failed, retrieved the key from secrets manager manually entered it on the directory services and badabing. it worked.

**Comment by nagmesh** [comment]
I found that for some reason SSM was not running Step 4 in the command execution of the doc **RadiusProxyBootstrapDocument**. Check the SSM steps if you have 3 then Duo MFA was not installed. Or check the number of instances in command history for **InstallAndConfigureDuoProxyServiceDocument**.
I removed all `precondition` steps in the SSM docs, destroyed and re-ran the template. It worked. Both in gov and commercial.
